tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,605 @@
+package tensorflow
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+const (
+	cOpsProtobufDefsPath = "/usr/local/tensorlow/ops.pbtxt"
+)
+
+// Graph Representation of the computation graph.
+type Graph struct {
+	def *pb.GraphDef
+
+	availableOps map[string]*pb.OpDef
+	constants    map[string]*Tensor
+	variables    map[string]*Tensor
+}
+
+// GraphNode Representation of one of the nodes of the TensorFlow Graph.
+// A node takes zero or more Tensors, performs some computation, and
+// produces zero or more Tensors.
+type GraphNode struct {
+	ref          *pb.NodeDef
+	def          *pb.NodeDef
+	outDataTypes map[string]DataType
+}
+
+// NewGraph Returns an initialized instance of the Graph struct.
+func NewGraph() *Graph {
+	return &Graph{
+		def:          new(pb.GraphDef),
+		availableOps: make(map[string]*pb.OpDef),
+		constants:    make(map[string]*Tensor),
+		variables:    make(map[string]*Tensor),
+	}
+}
+
+// NewGraphFromText Returns a new graph populated with the deserialization of
+// the provided graph string.
+func NewGraphFromText(graphStr string) (gr *Graph, err error) {
+	gr = NewGraph()
+	err = proto.UnmarshalText(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromFile Loads a Graph from the file on the specified path.
+func LoadGraphFromFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	gr = NewGraph()
+	err = proto.Unmarshal(graphStr, gr.def)
+
+	return
+}
+
+// LoadGraphFromTextFile Loads a Graph as plain text from the file on the specified
+// path.
+func LoadGraphFromTextFile(path string) (gr *Graph, err error) {
+	graphStr, err := ioutil.ReadFile(path)
+	if err != nil {
+		return
+	}
+
+	return NewGraphFromText(string(graphStr))
+}
+
+// Op Adds a new Node to the Graph with the specified operation, this function
+// could return an error if any of the mandatory attributes is not be present
+// or the value is not the expected for this attribute.
+func (gr *Graph) Op(opName string, name string, input []*GraphNode, device string, attrs map[string]interface{}) (node *GraphNode, err error) {
```

Comment:
I think this function would read a lot more cleanly internally if it were to use explicit return values.  For example, in the "operation not found case", this is both more concise and more clear:
  if .... !opFound {
         return gr, &ErrOperationNotFound{ op: opName }
  }

it makes it a little more clear that you're returning an error.  I think the repeated savings of this over the course of the function makes it worth the slight redundancy of having to note the 'gr' on every return.  This is a pretty big function, so it's reasonable to remind the reader what the return values are when they're 40 lines into it.  There are also some cases inside the code that already explicitly specify the return values, and the mix gets kind of confusing. 

It would also be worth checking in some of these cast checks if they read better with an integrated conditional, e.g., for this code:
dt, ok := v.(DataType)
 +				if !ok {
 +					return nil, &ErrInvalidAttrValue{
 +						operation:  opName,
 +						attribName: attr.Name,
 +					}
 +				}
 +				node.def.Attr[attr.Name] = &pb.AttrValue{
 +					Value: &pb.AttrValue_Type{
 +						Type: pb.DataType(dt),
 +					},
 +				}

Would this be more clear?
if dt, ok := v.(DataType); !ok {
   return nil, &ErrInvalidAttrValue{ ...}
} else {
   ...
}

I'm not 100% on this latter, so I leave it to you to see which reads better.